### DISCUSSION TITLE
[2.0] Mypy: Need type annotation for "async_session"

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#8837] [Usage Questions]
### Describe the bug

When using `async_sessionmaker` it seems that mypy is failing to infer the type if no `class_` argument is passed. The error disappears if specifying the argument.
```
app/_localdev/populate_db.py:73: error: Need type annotation for "async_session"  [var-annotated]
        async_session = reveal_type(async_sessionmaker(engine, expire_on_commit=False))
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
app/_localdev/populate_db.py:73: note: Revealed type is "sqlalchemy.ext.asyncio.session.async_sessionmaker[<nothing>]"
```

### Versions

- SQLAlchemy: 2b3
- Mypy: 0.991

**Comment by zzzeek** [comment]
this is a known beahvior of Mypy, where it can't know the type since async_sessionmaker has generic parameters.  the output tells you this:

```py
from typing import TYPE_CHECKING

from sqlalchemy.ext.asyncio import async_sessionmaker, AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine


e = create_async_engine("sqlite+aiosqlite://")

as_ = async_sessionmaker(e)

async_session = as_()

if TYPE_CHECKING:
    reveal_type(async_session)
```


```
test3.py:9: error: Need type annotation for "as_"  [var-annotated]
test3.py:14: note: Revealed type is "Any"
```

it should be annotated as:

```
from typing import TYPE_CHECKING

from sqlalchemy.ext.asyncio import async_sessionmaker, AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine


e = create_async_engine("sqlite+aiosqlite://")

as_: async_sessionmaker[AsyncSession] = async_sessionmaker(e)

async_session = as_()

if TYPE_CHECKING:
    reveal_type(async_session)
```

pylance doesnt have this issue and I am not aware of how it can be fixed in mypy, it's likely something people have reported.  this is how it's typed:

```py
_AS = TypeVar("_AS", bound="AsyncSession")


class async_sessionmaker(Generic[_AS]):
   ...

```

so mypy should know that the type is at least `async_sessionmaker[AsyncSession]` but it doesnt.

**Reply by CaselIT** [reply]
Also Python typing does not seem to have a way of specifying the default type of a generic. I guess bound comes close, but it does not seem to be used my mypy in this way

**Reply by Dreamsorcerer** [reply]
So, the default value is not part of the typing information, so I think it's perfectly fine for mypy to be ignoring it (e.g. in typeshed the stub would be annotated as `class_: Type[_AS] = ...`). Maybe it would be sensible for mypy to pick the upper bound as a reasonable type, but ideally we would use overloads to give it a concrete type.

I'll make a PR with a suggestion which should work correctly.

**Reply by zzzeek** [reply]
`async_sessionmaker()` is a class.  how do you apply "overloads" to that?

**Reply by Dreamsorcerer** [reply]
Roughly something like this: #8842

**Reply by CaselIT** [reply]
that does not seem to work

**Reply by zzzeek** [reply]
OK the `__init__()` method thing, I've had mixed results with that.  mypy or pylance has some problem with it.   both tools are constantly being released with new updates, so it's hard to track these things
